securedrop_export/export.py

```diff
@@ -0,0 +1,329 @@
+                json_config = json.loads(f.read())
+                self.export_method = json_config.get("device", None)
+                self.encryption_method = json_config.get("encryption_method", None)
+                self.encryption_key = json_config.get("encryption_key", None)
+        except Exception as e:
+            raise
+
+    def is_valid(self):
+        if self.export_method not in self.SUPPORTED_EXPORT_METHODS:
+            return False
+
+        if self.export_method == "disk":
+            if self.encryption_method not in self.SUPPORTED_ENCRYPTION_METHODS:
+                return False
+        return True
+
+
+class SDExport(object):
+
+    def __init__(self, archive):
+        self.device = DEVICE
+        self.mountpoint = MOUNTPOINT
+        self.encrypted_device = ENCRYPTED_DEVICE
+
+        self.printer_name = PRINTER_NAME
+        self.printer_wait_timeout = PRINTER_WAIT_TIMEOUT
+
+        self.brlaser_driver = BRLASER_DRIVER
+        self.brlaser_ppd = BRLASER_PPD
+        
+        self.archive = archive
+        self.submission_dirname = os.path.basename(self.archive).split(".")[0] 
+        self.target_dirname = "sd-export-{}".format(
+            datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
+        )
+        self.tmpdir = tempfile.mkdtemp()
+        self.archive_path = os.path.join(self.tmpdir, self.submission_dirname)
+
+
+    def exit_gracefully(self, msg, e=False):
+        """
+        Utility to print error messages, mostly used during debugging,
+        then exits successfully despite the error. Always exits 0,
+        since non-zero exit values will cause system to try alternative
+        solutions for mimetype handling, which we want to avoid.
+        """
+        sys.stderr.write(msg)
+        sys.stderr.write("\n")
+        if e:
+            try:
+                # If the file archive was extracted, delete before returning
+                if os.path.isdir(self.tmpdir):
+                    shutil.rmtree(self.tmpdir)
+                e_output = e.output
+            except Exception:
+                e_output = "<unknown exception>"
+            sys.stderr.write(e_output)
+            sys.stderr.write("\n")
+        # exit with 0 return code otherwise the os will attempt to open
+        # the file with another application
+        self.popup_message("Export error: {}".format(msg))
+        sys.exit(0)
+
+
+    def popup_message(self, msg):
+        try:
+            subprocess.check_call([
+                "notify-send",
+                "--expire-time", "3000",
+                "--icon", "/usr/share/securedrop/icons/sd-logo.png",
+                "SecureDrop: {}".format(msg)
+            ])
+        except subprocess.CalledProcessError as e:
+            msg = "Error sending notification:"
+            self.exit_gracefully(msg, e=e)
+
+
+    def extract_tarball(self):
+        try:
+            with tarfile.open(self.archive) as tar:
+                tar.extractall(self.archive_path)
```

Comment:
tar will create a folder with the name of the archive, to the script would the temporary folder directly: 
changing this line to `tar.extractall(self.self.tmpdir)` resolves for me locally

Reply:
This was originally self.tmpdir (or SUBMISSION_TMPDIR in the single-file version of the script), but that was breaking metadata extraction for me for submission bundles named `<foo>.sd_export` where `<foo>` was any string other than null. It could also be updated in the Metadata class I guess.

Reply:
OK, updated both here and in __main__() where Metadata gets initted. fixed.